pkg/controller/util/endpoint/controller_utils_test.go

```diff
@@ -224,3 +231,168 @@ func TestShouldPodBeInEndpoints(t *testing.T) {
+					"app": fmt.Sprintf("test-%d", i),
+				},
+			},
+		}
+		fakeInformerFactory.Core().V1().Services().Informer().GetStore().Add(service)
+	}
+	var pods []*v1.Pod
+	for i := 0; i < 5; i++ {
+		pod := &v1.Pod{
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace: "test",
+				Name:      fmt.Sprintf("test-pod-%d", i),
+				Labels: map[string]string{
+					"app":   fmt.Sprintf("test-%d", i),
+					"label": fmt.Sprintf("label-%d", i),
+				},
+			},
+		}
+		pods = append(pods, pod)
+	}
+
+	cache := NewServiceSelectorCache()
+	tests := []struct {
+		name   string
+		pod    *v1.Pod
+		expect sets.String
+	}{
+		{
+			name:   "get servicesMemberships for pod-0",
+			pod:    pods[0],
+			expect: sets.String{"test/service-0": sets.Empty{}},
+		},
+		{
+			name:   "get servicesMemberships for pod-1",
+			pod:    pods[1],
+			expect: sets.String{"test/service-1": sets.Empty{}},
+		},
+		{
+			name:   "get servicesMemberships for pod-2",
+			pod:    pods[2],
+			expect: sets.String{"test/service-2": sets.Empty{}},
+		},
+		{
+			name:   "get servicesMemberships for pod-3",
+			pod:    pods[3],
+			expect: sets.String{},
+		},
+		{
+			name:   "get servicesMemberships for pod-4",
+			pod:    pods[4],
+			expect: sets.String{},
+		},
+	}
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			if services, err := cache.GetPodServiceMemberships(fakeInformerFactory.Core().V1().Services().Lister(), test.pod); err != nil || !services.Equal(test.expect) {
+				t.Errorf("%s fail with error %v, get %v but expect %v", test.name, err, services, test.expect)
+			}
+		})
+	}
+}
+
+func TestServiceSelectorCache_Update(t *testing.T) {
+	var selectors []labels.Selector
+	for i := 0; i < 5; i++ {
+		selector := labels.Set(map[string]string{"app": fmt.Sprintf("test-%d", i)}).AsSelectorPreValidated()
+		selectors = append(selectors, selector)
+	}
+	tests := []struct {
+		name   string
+		key    string
+		cache  *ServiceSelectorCache
+		update map[string]string
+		expect labels.Selector
+	}{
+		{
+			name:   "add test/service-0",
+			key:    "test/service-0",
+			cache:  generateSelectorCache(map[string]labels.Selector{}),
+			update: map[string]string{"app": "test-0"},
+			expect: selectors[0],
+		},
+		{
+			name:   "add test/service-1",
+			key:    "test/service-1",
+			cache:  generateSelectorCache(map[string]labels.Selector{"test/service-0": selectors[0]}),
+			update: map[string]string{"app": "test-1"},
+			expect: selectors[1],
+		},
+		{
+			name:   "update test/service-2",
+			key:    "test/service-2",
+			cache:  generateSelectorCache(map[string]labels.Selector{"test/service-2": selectors[2]}),
+			update: map[string]string{"app": "test-0"},
+			expect: selectors[0],
+		},
+	}
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			selector := test.cache.Update(test.key, test.update)
+			if !reflect.DeepEqual(selector, test.expect) {
+				t.Errorf("%s failed with selector %v but expect %v", test.name, selector, test.expect)
```

Comment:
Small thing: I'd reword this error message to follow the expected + got pattern I've seen elsewhere in the kubernetes codebase, something like:

```
Expected selector: %v, got: %v
```

In all these cases since you're in a `t.Run` block you don't actually need to write the test name out again.